### ISSUE TITLE
str_replace with a polymorphic subject returns string

### ISSUE THREAD

**Issue by egst** [vimeo/psalm#7625]
The return type of `str_replace` is deduced to be `string` when I pass it a `$subject` (the third parameter) of type `string | array<string>`. A `$subject` of type `string` leads to the `string` return type, while `array<float | int | string>` leads to `array<string>`, which is correct. It only incorrectly assumes `string` when the `$subject` is of a union type.

https://psalm.dev/r/48f8c3f020

**Comment by egst** [comment]
When I wrap the `str_replace` function in another one with the same type annotations that are in Psalm stub files (`stubs/CoreGenericFunctions.phpstub`), it all works ok. So it seems that all of the mechanisms used in this annotation work perfectly well. Maybe the check doesn't consider the correct stub file for some reason or something like that?

https://psalm.dev/r/f4be0e3119

**Comment by AndrolGenhald** [comment]
The return type is being overridden by `Psalm\Internal\Provider\ReturnTypeProvider\StrReplaceReturnTypeProvider`. Either it's not necessary anymore and can be removed, or it's not accurate and needs to be fixed.

**Comment by orklah** [comment]
Yeah, clearly a wrong behaviour of StrReplaceReturnTypeProvider. When the param is an array without a string, it returns array:
https://github.com/vimeo/psalm/blob/b924032850df54dd05ffae409a709820fd6e688e/src/Psalm/Internal/Provider/ReturnTypeProvider/StrReplaceReturnTypeProvider.php#L47

In all other cases, it returns a string: https://github.com/vimeo/psalm/blob/b924032850df54dd05ffae409a709820fd6e688e/src/Psalm/Internal/Provider/ReturnTypeProvider/StrReplaceReturnTypeProvider.php#L50

This should be easy to fix.

Callmaps definition are never removed when we have a return type provider. I think stubs are also overrided, but it seems pointless and error prone to keep stubs in this case, so I think it should be removed

**Comment by AndrolGenhald** [comment]
> it seems pointless and error prone to keep stubs in this case,

The stub also has `@psalm-flow` for taint analysis, I don't think that's doable in the callmap. At least a comment could be added to it that the return type is overridden by a provider, and I think if the stub is kept the callmap definition can be removed?

**Comment by orklah** [comment]
we never remove anything from the callmap. Not sure if it's by design or just because we don't need to.

I'm also not sure if @psalm-flow has any effect when we have a return type provider overriding the stub...
